### PR TITLE
i#3129 raw2trace perf: factored trace_converter_t out of raw2trace_t

### PR DESCRIPTION
**Pull request by mtrofin**
Introducing trace_converter_t, a template factoring out of raw2trace_t the functionality necessary for building custom trace converters.

Issue #3129

### CODE REVIEW

File: clients/drcachesim/tracer/raw2trace.h [hunk 1]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
 /**
- * The raw2trace class converts the raw offline trace format to the format
- * expected by analysis tools.  It requires access to the binary files for the
- * libraries and executable that were present during tracing.
+ * trace_converter_t is a reusable component that encapsulates raw trace conversion.
+ *
+ * Conversion happens from a data source abstracted by the type parameter T. We assume
+ * trace entry records belonging to a thread follow together, however, the way such
+ * records are obtained is left for implementers to decide. For example, all thread
+ * records of a thread may be in the same file; or records from multiple threads  might be
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 2]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ * interleaved accross multiple files.
+ *
+ * trace_converter_t expects to be derived by its type template T which should provide the
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 3]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ * following APIs:
+ *
+ * <UL> <LI>const offline_entry_t *get_next_entry()
+ *
+ * Point to the next offline entry_t. There is no assumption of the underlying source
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 4]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ * of the data, and trace_converter_t will not attempt to dereference past the provided
+ * pointer.</LI>
+ *
+ * <LI>void unread_last_entry()
+ *
+ * Ensure that the next call to get_next_entry() re-reads the last value.</LI>
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: Ya - it looks like a dot, the function signature, then a newline and the documentation for that function.

We could do some other formatting, maybe more compact - like bold out or italicize the function name and then continue with the doc?

File: clients/drcachesim/tracer/raw2trace.h [hunk 5]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+#define INVALID_THREAD_ID 0
+
 /**
- * The raw2trace class converts the raw offline trace format to the format
- * expected by analysis tools.  It requires access to the binary files for the
- * libraries and executable that were present during tracing.
+ * trace_converter_t is a reusable component that encapsulates raw trace conversion.
+ *
+ * Conversion happens from a data source abstracted by the type parameter T. We assume
+ * trace entry records belonging to a thread follow together, however, the way such
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 6]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ * of the data, and trace_converter_t will not attempt to dereference past the provided
+ * pointer.</LI>
+ *
+ * <LI>void unread_last_entry()
+ *
+ * Ensure that the next call to get_next_entry() re-reads the last value.</LI>
+ *
+ * <LI>trace_entry_t *get_write_buffer()
+ *
+ * Return a writable buffer guaranteed to be at least WRITE_BUFFER_SIZE large.
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: ah! I kept looking for docs on how to ref fields, including \ref typename::field, and failed. There's one more place this should be needed, too.

File: clients/drcachesim/tracer/raw2trace.h [hunk 7]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ *  get_write_buffer() may reuse the same buffer after write or write_delayed_branches is
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 8]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ * <LI>bool write(const trace_entry_t *start, const trace_entry_t *end)
+ *
+ * Writes the converted traces between start and end, where end is past the last
+ * item to write. Both start and end are assumed to be pointers inside a buffer
+ * returned by get_write_buffer().</LI>
+ *
+ * <LI>std::string write_delayed_branches(const trace_entry_t *start, const trace_entry_t
+ * *end)
+ *
+ * Simliar to write(), but treat the provided traces as delayed branches: if they
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: thanks. done.

File: clients/drcachesim/tracer/raw2trace.h [hunk 9]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ * <LI>std::string on_thread_end()
+ *
+ * Callback notifying the currently-processed thread has exited. trace_converter_t
+ * extenders are expected to track record metadata themselves. trace_converter_t offers
+ * APIs for extracting that metadata.</LI>
+ *
+ * <LI>void log(uint level, const char *fmt, ...)
+ *
+ * Implementers are given the opportunity to implement their own logging. The level
+ * parameter represents severity, the lower the level, the higher the severity.</LI>
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 10]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+ *
+ * <LI>const instr_summary_t *get_instr_summary(uint64 modx, uint64 modoffs, INOUT app_pc
+ * *pc, app_pc orig)
+ *
+ * Return the #instr_summary_t representation of the instruction at *pc,
+ * updating the value at pc to the PC of the next instruction. It is assumed the app
+ * binaries have allready been loaded using #module_mapper_t, and the values at *pc point
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 11]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
-     * iterates over the list of modules.  This is provided to give the user a method
-     * for iterating modules in the presence of the custom field used by drmemtrace
-     * that prevents direct use of drmodtrack_offline_read().
-     * On success, calls the \p process_cb function passed to handle_custom_data()
-     * for every module in the list, and returns an empty string at the end.
-     * Returns a non-empty error message on failure.
-     *
-     * \deprecated #module_mapper_t should be used instead.
+     * Construct a new trace_converter_t object. If a nullptr dcontext_in is passed,
+     * creates a new DR context va dr_standalone_init()
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h [hunk 12]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
-     * Performs the first step of do_conversion() without further action: parses and
-     * iterates over the list of modules.  This is provided to give the user a method
-     * for iterating modules in the presence of the custom field used by drmemtrace
-     * that prevents direct use of drmodtrack_offline_read().
-     * On success, calls the \p process_cb function passed to handle_custom_data()
-     * for every module in the list, and returns an empty string at the end.
-     * Returns a non-empty error message on failure.
-     *
-     * \deprecated #module_mapper_t should be used instead.
+     * Construct a new trace_converter_t object. If a nullptr dcontext_in is passed,
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: OK - I intentionally opted against that, thinking that the reader is already in the trace_converter_t context (page). I'll add though, can't hurt.

File: clients/drcachesim/tracer/raw2trace.h [hunk 13]
```diff
@@ -356,119 +356,94 @@ struct trace_header_t {
...
+    trace_converter_t(void *dcontext_in)
+        : dcontext(dcontext_in == nullptr ? dr_standalone_init() : dcontext_in)
+    {
+    }
 
     /**
-     * Performs the conversion from raw data to finished trace files.
-     * Returns a non-empty error message on failure.
+     * Convert starting from in_entry, and reading more entries as required.
+     * sets end_of_record to true if processing hit the end of a record.
```
[the remark this answers is not in the document]
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

File: clients/drcachesim/tracer/raw2trace.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.h]: FTR I mentioned the `#` prefix in several prior reviews

File: clients/drcachesim/tracer/raw2trace.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.h]: This is a little strange to have these functions listed here.  Normally you would put the doxygen comments in front of each function declaration in the header file, which will add a proper entry.  In fact it looks like you don't have any comment at all there, which means there's presumably no link added for the `()` refs?  I would suggest moving all this to proper doxygen comments in front of the respective functions and just put links here (i.e., just list the functions with `()` to get links).

You can use `@param[in] ...` etc. to make nicer docs for each function.

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.h]: I guess to do that you'd have to declare an interface class for T, which maybe is not a bad idea, to get the compiler checking that is has these required functions.

File: clients/drcachesim/tracer/raw2trace.h [hunk 14]
```diff
@@ -360,21 +360,21 @@ struct trace_header_t {
...
- * Conversion happens from a data source abstracted by the type parameter T. We assume
- * trace entry records belonging to a thread follow together, however, the way such
- * records are obtained is left for implementers to decide. For example, all thread
- * records of a thread may be in the same file; or records from multiple threads  might be
- * interleaved accross multiple files.
+ * Conversion happens from a data source abstracted by the type parameter T. We make no
+ * assumption about how thread buffers are organized. We do assume the internal
+ * composition of thread buffers is "as written" by the thread. For example, all thread
+ * buffers belonging to different threads may be in a separate files; or buffers may be
+ * co-located in one large file, or spread accross multiple, mix-thread files.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.h]: nit: s/mix/mixed/

File: clients/drcachesim/tracer/raw2trace.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tracer/raw2trace.h]: To do what my prior comments are trying to say would require putting these comments under the raw2trace_t implementation, which would require `raw2trace_t::` in front of them all here to get a link: I don't think there's cleaner way.  Seems ok to leave it as is.

File: clients/drcachesim/tracer/raw2trace.h (hunk 14 above)
Comment (mtrofin) [reply clients/drcachesim/tracer/raw2trace.h]: done

**Pull request merged**